// prelude/lib/object.js
// Generated by LiveScript 1.4.0
'use strict';
var curry, array, ref$, isString, isObject, isArray, ObjHasOwnProperty, slice$ = [].slice;
curry = require('./curry');
array = require('./array');
ref$ = require('./type'), isString = ref$.isString, isObject = ref$.isObject, isArray = ref$.isArray;
ObjHasOwnProperty = Object.prototype.hasOwnProperty;
exports.empty = function(obj){
  var k;
  for (k in obj) {
    return false;
  }
  return true;
};
exports.has = curry(function(key, obj){
  return ObjHasOwnProperty.call(obj, key);
});
exports.includes = curry(function(elem, obj){
  var i$, v;
  for (i$ in obj) {
    v = obj[i$];
    if (v === elem) {
      return true;
    }
  }
  return false;
});
exports.contains = exports.includes;
exports.keys = function(obj){
  var k, results$ = [];
  for (k in obj) {
    results$.push(k);
  }
  return results$;
};
exports.values = function(obj){
  var i$, v, results$ = [];
  for (i$ in obj) {
    v = obj[i$];
    results$.push(v);
  }
  return results$;
};
exports.clone = function(obj){
  return exports.deepMerge({}, obj);
};
exports.flatten = curry(function(delimiter, obj){
  var result;
  result = {};
  (function(){
    function parse(obj, parentKey){
      var key, child;
      parentKey == null && (parentKey = '');
      for (key in obj) {
        child = obj[key];
        if (parentKey && isString(delimiter)) {
          key = parentKey + "" + delimiter + key;
        }
        if (isObject(child) && Object.keys(child).length > 0) {
          parse(child, key);
        } else {
          result[key] = child;
        }
      }
    }
    return parse;
  }())(
  obj);
  return result;
});
exports.explode = curry(function(delimiter, obj){
  var result;
  result = {};
  (function(){
    function parse(obj, parentKey){
      var flatKey, value, ref, keys, len, i$, len$, i, k;
      parentKey == null && (parentKey = '');
      for (flatKey in obj) {
        value = obj[flatKey];
        ref = result;
        keys = flatKey.split(delimiter);
        len = keys.length - 1;
        for (i$ = 0, len$ = keys.length; i$ < len$; ++i$) {
          i = i$;
          k = keys[i$];
          if (i < len) {
            if (!isObject(ref[k])) {
              ref[k] = {};
            }
            ref = ref[k];
            continue;
          }
          ref[k] = value;
        }
      }
    }
    return parse;
  }())(
  obj);
  return result;
});
exports.each = curry(function(f, obj){
  var k, v;
  for (k in obj) {
    v = obj[k];
    f(v, k);
  }
  return obj;
});
exports.map = curry(function(f, obj){
  var k, v, resultObj$ = {};
  for (k in obj) {
    v = obj[k];
    resultObj$[k] = f(v, k);
  }
  return resultObj$;
});
exports.filter = curry(function(f, obj){
  var k, v, resultObj$ = {};
  for (k in obj) {
    v = obj[k];
    if (f(v, k)) {
      resultObj$[k] = v;
    }
  }
  return resultObj$;
});
exports.every = curry(function(f, obj){
  var k, v;
  for (k in obj) {
    v = obj[k];
    if (f(v, k) === false) {
      return false;
    }
  }
  return true;
});
exports.some = curry(function(f, obj){
  var k, v;
  for (k in obj) {
    v = obj[k];
    if (f(v, k) === true) {
      return true;
    }
  }
  return false;
});
exports.partition = curry(function(f, obj){
  var passed, failed, k, v;
  passed = {};
  failed = {};
  for (k in obj) {
    v = obj[k];
    (f(v, k) ? passed : failed)[k] = v;
  }
  return [passed, failed];
});
exports.keyOf = curry(function(elem, obj){
  var k, v;
  for (k in obj) {
    v = obj[k];
    if (v === elem) {
      return k;
    }
  }
});
exports.keysOf = curry(function(elem, obj){
  var k, v, results$ = [];
  for (k in obj) {
    v = obj[k];
    if (v === elem) {
      results$.push(k);
    }
  }
  return results$;
});
exports.findKey = curry(function(f, obj){
  var k, v;
  for (k in obj) {
    v = obj[k];
    if (f(v, k)) {
      return k;
    }
  }
});
exports.findKeys = curry(function(f, obj){
  var k, v, results$ = [];
  for (k in obj) {
    v = obj[k];
    if (f(v, k)) {
      results$.push(k);
    }
  }
  return results$;
});
exports.fromPairs = function(xs){
  var i$, len$, x, resultObj$ = {};
  for (i$ = 0, len$ = xs.length; i$ < len$; ++i$) {
    x = xs[i$];
    resultObj$[x[0]] = x[1];
  }
  return resultObj$;
};
exports.toPairs = function(obj){
  var key, value, results$ = [];
  for (key in obj) {
    value = obj[key];
    results$.push([key, value]);
  }
  return results$;
};
exports.fill = curry(1, function(init){
  var sources, res, i$, len$, src, k;
  sources = slice$.call(arguments, 1);
  res = {};
  for (i$ = 0, len$ = sources.length; i$ < len$; ++i$) {
    src = sources[i$];
    for (k in init) {
      if (src[k] != null) {
        res[k] = src[k];
      }
    }
  }
  return res;
});
exports.deepFill = curry(1, function(init){
  var sources, res, i$, len$, src, key, own$ = {}.hasOwnProperty;
  sources = slice$.call(arguments, 1);
  res = {};
  for (i$ = 0, len$ = sources.length; i$ < len$; ++i$) {
    src = sources[i$];
    for (key in init) if (own$.call(init, key)) {
      if (src[key] != null) {
        if (isObject(src[key]) && isObject(init[key])) {
          res[key] = exports.deepFill(init[key], src[key]);
        } else {
          res[key] = src[key];
        }
      }
    }
  }
  return res;
});
exports.merge = curry(1, function(){
  var sources, dest, i$, len$, src, key, val, own$ = {}.hasOwnProperty;
  sources = slice$.call(arguments);
  dest = {};
  for (i$ = 0, len$ = sources.length; i$ < len$; ++i$) {
    src = sources[i$];
    for (key in src) if (own$.call(src, key)) {
      val = src[key];
      dest[key] = val;
    }
  }
  return dest;
});
exports.deepMerge = curry(1, function(){
  var sources, dest, i$, len$, src, key, value;
  sources = slice$.call(arguments);
  dest = {};
  for (i$ = 0, len$ = sources.length; i$ < len$; ++i$) {
    src = sources[i$];
    for (key in src) {
      value = src[key];
      if (isObject(value) && isObject(dest[key])) {
        dest[key] = exports.deepMerge({}, dest[key], value);
      } else if (isArray(value)) {
        dest[key] = array.clone(value);
      } else {
        dest[key] = value;
      }
    }
  }
  return dest;
});
exports.freeze = function(obj){
  return Object.freeze(obj);
};
exports.deepFreeze = function(obj){
  var key, value;
  if (!Object.isFrozen(obj)) {
    Object.freeze(obj);
  }
  for (key in obj) {
    value = obj[key];
    if (ObjHasOwnProperty.call(obj, key) && isObject(value)) {
      exports.deepFreeze(value);
    }
  }
  return obj;
};
exports.toJSON = function(obj, indentBy){
  indentBy == null && (indentBy = 0);
  return JSON.stringify(obj, void 8, indentBy);
};
exports.fromJSON = function(obj, fn){
  var res, err;
  try {
    res = JSON.parse(obj);
  } catch (e$) {
    err = e$;
  }
  if (typeof fn === 'function') {
    fn(err, res);
  }
  return res;
};
exports.fromJSONUnsafe = function(obj){
  return JSON.parse(obj);
};
exports.definePublic = curry(1, function(obj, key, value){
  var k, v;
  if (isObject(key)) {
    for (k in key) {
      v = key[k];
      exports.definePublic(obj, k, v);
    }
    return obj;
  } else {
    return Object.defineProperty(obj, key, {
      value: value,
      enumerable: true,
      writable: true,
      configurable: true
    });
  }
});
exports.definePrivate = curry(1, function(obj, key, value){
  var k, v;
  if (isObject(key)) {
    for (k in key) {
      v = key[k];
      exports.definePrivate(obj, k, v);
    }
    return obj;
  } else {
    return Object.defineProperty(obj, key, {
      value: value,
      enumerable: false,
      writable: true,
      configurable: true
    });
  }
});
exports.defineStatic = curry(1, function(obj, key, value){
  var k, v;
  if (isObject(key)) {
    for (k in key) {
      v = key[k];
      exports.defineStatic(obj, k, v);
    }
    return obj;
  } else {
    return Object.defineProperty(obj, key, {
      value: value,
      enumerable: true,
      writable: false,
      configurable: true
    });
  }
});
exports.defineMeta = curry(1, function(obj, key, value){
  var k, v;
  if (isObject(key)) {
    for (k in key) {
      v = key[k];
      exports.defineMeta(obj, k, v);
    }
    return obj;
  } else {
    return Object.defineProperty(obj, key, {
      value: value,
      enumerable: false,
      writable: false,
      configurable: true
    });
  }
});
