// prelude/lib/string.js
// Generated by LiveScript 1.4.0
'use strict';
var curry, array, TRIM_LEFT, TRIM_RIGHT, TRIM_BOTH;
curry = require('./curry');
array = require('./array');
TRIM_LEFT = new RegExp('^[s\uFEFF\xA0]+');
TRIM_RIGHT = new RegExp('[s\uFEFF\xA0]+$');
TRIM_BOTH = new RegExp('^[s\uFEFF\xA0]+|[s\uFEFF\xA0]+$', 'g');
exports.empty = function(str){
  return str.length === 0;
};
exports.includes = function(part, str){
  return str.indexOf(part) !== -1;
};
exports.contains = exports.includes;
exports.startsWith = curry(1, function(search, str){
  return search === str.slice(0, search.length);
});
exports.endsWith = curry(1, function(search, str){
  return search === str.slice(str.length - search.length);
});
exports.trim = function(str){
  return str.replace(TRIM_BOTH, '');
};
exports.trimLeft = function(str){
  return str.replace(TRIM_LEFT, '');
};
exports.trimRight = function(str){
  return str.replace(TRIM_RIGHT, '');
};
exports.repeat = curry(function(n, str){
  return (function(){
    var i$, to$, results$ = [];
    for (i$ = 0, to$ = n; i$ < to$; ++i$) {
      results$.push(str);
    }
    return results$;
  }()).join('');
});
exports.reverse = function(str){
  var i, len, result;
  i = 0;
  len = str.length;
  result = new Array(len);
  while (len !== 0) {
    result[--len] = str.charAt(i++);
  }
  return result.join('');
};
exports.capitalize = function(str){
  return str.charAt(0).toUpperCase() + str.slice(1);
};
exports.capitalizeSentence = function(str){
  return str.replace(/(^.|\s.)/g, function(arg$, c){
    return (c != null ? c : '').toUpperCase();
  });
};
exports.decapitalize = function(str){
  return str.charAt(0).toLowerCase() + str.slice(1);
};
exports.decapitalizeSentence = function(str){
  return str.replace(/(^.|\s.)/g, function(arg$, c){
    return (c != null ? c : '').toLowerCase();
  });
};
exports.camelize = function(str){
  return str.replace(/[-_]+(.)?/g, function(arg$, c){
    return (c != null ? c : '').toUpperCase();
  });
};
exports.dasherize = function(str){
  return str.replace(/([^-A-Z])([A-Z]+)/g, function(arg$, lower, upper){
    upper = upper.length > 1
      ? upper
      : upper.toLowerCase();
    return lower + '-' + upper;
  }).replace(/^([A-Z]+)/, function(arg$, upper){
    if (upper.length > 1) {
      return upper + '-';
    } else {
      return upper.toLowerCase();
    }
  });
};
