// prelude/lib/array.js
// Generated by LiveScript 1.4.0
'use strict';
var curry, number, object, ref$, isArray, isObject, inRange, apply, slice$ = [].slice;
curry = require('./curry');
number = require('./number');
object = require('./object');
ref$ = require('./type'), isArray = ref$.isArray, isObject = ref$.isObject, inRange = ref$.inRange;
apply = require('./func').apply;
exports.empty = function(xs){
  return xs.length === 0;
};
exports.has = curry(function(i, xs){
  return i >= 0 && i <= xs.length - 1;
});
exports.includes = curry(function(y, xs){
  var i$, len$, x;
  for (i$ = 0, len$ = xs.length; i$ < len$; ++i$) {
    x = xs[i$];
    if (x === y) {
      return true;
    }
  }
  return false;
});
exports.contains = exports.includes;
exports.clone = function(xs){
  var copy, i$, len$, i, value;
  copy = [];
  for (i$ = 0, len$ = xs.length; i$ < len$; ++i$) {
    i = i$;
    value = xs[i$];
    if (isObject(value)) {
      copy[i] = object.deepMerge({}, value);
    } else if (isArray(value)) {
      copy[i] = exports.clone(value);
    } else {
      copy[i] = value;
    }
  }
  return copy;
};
exports.first = exports.head = function(xs){
  return xs[0];
};
exports.rest = exports.tail = function(xs){
  var i$, len$, i, x, results$ = [];
  for (i$ = 0, len$ = xs.length; i$ < len$; ++i$) {
    i = i$;
    x = xs[i$];
    if (i > 0) {
      results$.push(x);
    }
  }
  return results$;
};
exports.last = function(xs){
  return xs[xs.length - 1];
};
exports.initial = function(xs){
  var init, i$, len$, i, x, results$ = [];
  if (!(init = xs.length)) {
    return [];
  }
  --init;
  for (i$ = 0, len$ = xs.length; i$ < len$; ++i$) {
    i = i$;
    x = xs[i$];
    if (i < init) {
      results$.push(x);
    }
  }
  return results$;
};
exports.slice = curry(function(a, b, xs){
  return xs.slice(a, b);
});
exports.splice = curry(2, function(pos, len, xs){
  var i$, len$, i, x, results$ = [];
  for (i$ = 0, len$ = xs.length; i$ < len$; ++i$) {
    i = i$;
    x = xs[i$];
    if (!inRange(pos, pos + len - 1, i)) {
      results$.push(x);
    }
  }
  return results$;
});
exports.concat = curry(1, function(){
  var xxs, result, i$, len$, xs, j$, len1$, x;
  xxs = slice$.call(arguments);
  result = [];
  for (i$ = 0, len$ = xxs.length; i$ < len$; ++i$) {
    xs = xxs[i$];
    for (j$ = 0, len1$ = xs.length; j$ < len1$; ++j$) {
      x = xs[j$];
      result.push(x);
    }
  }
  return result;
});
exports.remove = curry(1, function(y, xs){
  var i$, len$, x, results$ = [];
  for (i$ = 0, len$ = xs.length; i$ < len$; ++i$) {
    x = xs[i$];
    if (x !== y) {
      results$.push(x);
    }
  }
  return results$;
});
exports.removeOne = curry(1, function(y, xs){
  var result, foundOne, i$, len$, x;
  result = [];
  foundOne = false;
  for (i$ = 0, len$ = xs.length; i$ < len$; ++i$) {
    x = xs[i$];
    if (x === y && !foundOne) {
      foundOne = true;
    } else {
      result.push(x);
    }
  }
  return result;
});
exports.flatten = function flatten(xs){
  var result, i$, len$, x;
  result = [];
  for (i$ = 0, len$ = xs.length; i$ < len$; ++i$) {
    x = xs[i$];
    if (isArray(x)) {
      result.push.apply(result, flatten(x));
    } else {
      result.push(x);
    }
  }
  return result;
};
exports.reverse = function(xs){
  var i, len, result;
  i = 0;
  len = xs.length;
  result = new Array(len);
  while (len !== 0) {
    result[--len] = xs[i++];
  }
  return result;
};
exports.each = curry(function(f, xs){
  var i$, len$, i, x;
  for (i$ = 0, len$ = xs.length; i$ < len$; ++i$) {
    i = i$;
    x = xs[i$];
    f(x, i);
  }
  return xs;
});
exports.map = curry(function(f, xs){
  var i$, len$, i, x, results$ = [];
  for (i$ = 0, len$ = xs.length; i$ < len$; ++i$) {
    i = i$;
    x = xs[i$];
    results$.push(f(x, i));
  }
  return results$;
});
exports.find = exports.filter = curry(function(f, xs){
  var i$, len$, i, x, results$ = [];
  for (i$ = 0, len$ = xs.length; i$ < len$; ++i$) {
    i = i$;
    x = xs[i$];
    if (f(x, i)) {
      results$.push(x);
    }
  }
  return results$;
});
exports.findOne = curry(function(f, xs){
  var i$, len$, i, x;
  for (i$ = 0, len$ = xs.length; i$ < len$; ++i$) {
    i = i$;
    x = xs[i$];
    if (f(x, i)) {
      return x;
    }
  }
});
exports.shuffle = function(xs){
  var result, i$, len$, i, x, r;
  result = new Array(xs.length);
  for (i$ = 0, len$ = xs.length; i$ < len$; ++i$) {
    i = i$;
    x = xs[i$];
    r = number.random(i);
    if (r !== i) {
      result[i] = result[r];
    }
    result[r] = xs[i];
  }
  return result;
};
exports.every = curry(function(f, xs){
  var i$, len$, i, x;
  for (i$ = 0, len$ = xs.length; i$ < len$; ++i$) {
    i = i$;
    x = xs[i$];
    if (f(x, i) === false) {
      return false;
    }
  }
  return true;
});
exports.some = curry(function(f, xs){
  var i$, len$, i, x;
  for (i$ = 0, len$ = xs.length; i$ < len$; ++i$) {
    i = i$;
    x = xs[i$];
    if (f(x, i) === true) {
      return true;
    }
  }
  return false;
});
exports.zip = curry(1, function(){
  var args, minLength, i$, len$, xs, ref$, i, lresult$, j$, results$ = [];
  args = slice$.call(arguments);
  minLength = 9e9;
  for (i$ = 0, len$ = args.length; i$ < len$; ++i$) {
    xs = args[i$];
    minLength <= (ref$ = xs.length) || (minLength = ref$);
  }
  for (i$ = 0; i$ < minLength; ++i$) {
    i = i$;
    lresult$ = [];
    for (j$ = 0, len$ = args.length; j$ < len$; ++j$) {
      xs = args[j$];
      lresult$.push(xs[i]);
    }
    results$.push(lresult$);
  }
  return results$;
});
exports.zipWith = curry(2, function(f){
  var args, minLength, i$, len$, xs, ref$, i, results$ = [];
  args = slice$.call(arguments, 1);
  minLength = 9e9;
  for (i$ = 0, len$ = args.length; i$ < len$; ++i$) {
    xs = args[i$];
    minLength <= (ref$ = xs.length) || (minLength = ref$);
  }
  for (i$ = 0; i$ < minLength; ++i$) {
    i = i$;
    results$.push(apply(f, (fn$())));
  }
  return results$;
  function fn$(){
    var i$, ref$, len$, results$ = [];
    for (i$ = 0, len$ = (ref$ = args).length; i$ < len$; ++i$) {
      xs = ref$[i$];
      results$.push(xs[i]);
    }
    return results$;
  }
});
exports.partition = curry(function(f, xs){
  var passed, failed, i$, len$, x;
  passed = [];
  failed = [];
  for (i$ = 0, len$ = xs.length; i$ < len$; ++i$) {
    x = xs[i$];
    (f(x) ? passed : failed).push(x);
  }
  return [passed, failed];
});
exports.unique = function(xs){
  var result, i$, len$, x;
  result = [];
  for (i$ = 0, len$ = xs.length; i$ < len$; ++i$) {
    x = xs[i$];
    if (!in$(x, result)) {
      result.push(x);
    }
  }
  return result;
};
exports.uniqueBy = curry(function(f, xs){
  var seen, i$, len$, x, val, results$ = [];
  seen = [];
  for (i$ = 0, len$ = xs.length; i$ < len$; ++i$) {
    x = xs[i$];
    val = f(x);
    if (in$(val, seen)) {
      continue;
    }
    seen.push(val);
    results$.push(x);
  }
  return results$;
});
exports.difference = curry(1, function(xs){
  var yss, result, i$, len$, x, j$, len1$, ys;
  yss = slice$.call(arguments, 1);
  result = [];
  diff: for (i$ = 0, len$ = xs.length; i$ < len$; ++i$) {
    x = xs[i$];
    for (j$ = 0, len1$ = yss.length; j$ < len1$; ++j$) {
      ys = yss[j$];
      if (in$(x, ys)) {
        continue diff;
      }
    }
    result.push(x);
  }
  return result;
});
exports.intersection = curry(1, function(xs){
  var yss, result, i$, len$, x, j$, len1$, ys;
  yss = slice$.call(arguments, 1);
  result = [];
  outer: for (i$ = 0, len$ = xs.length; i$ < len$; ++i$) {
    x = xs[i$];
    for (j$ = 0, len1$ = yss.length; j$ < len1$; ++j$) {
      ys = yss[j$];
      if (!in$(x, ys)) {
        continue outer;
      }
    }
    result.push(x);
  }
  return result;
});
exports.union = curry(1, function(){
  var xss, result, i$, len$, xs, j$, len1$, x;
  xss = slice$.call(arguments);
  result = [];
  for (i$ = 0, len$ = xss.length; i$ < len$; ++i$) {
    xs = xss[i$];
    for (j$ = 0, len1$ = xs.length; j$ < len1$; ++j$) {
      x = xs[j$];
      if (!in$(x, result)) {
        result.push(x);
      }
    }
  }
  return result;
});
exports.sortBy = curry(function(f, xs){
  return xs.concat().sort(function(x, y){
    var a, b;
    a = f(x);
    b = f(y);
    if (a > b) {
      return 1;
    } else if (a < b) {
      return -1;
    } else {
      return 0;
    }
  });
});
exports.countBy = curry(function(f, xs){
  var result, i$, len$, x, key;
  result = {};
  for (i$ = 0, len$ = xs.length; i$ < len$; ++i$) {
    x = xs[i$];
    key = f(x);
    if (key in result) {
      result[key] += 1;
    } else {
      result[key] = 1;
    }
  }
  return result;
});
exports.groupBy = curry(function(f, xs){
  var result, i$, len$, x, key;
  result = {};
  for (i$ = 0, len$ = xs.length; i$ < len$; ++i$) {
    x = xs[i$];
    key = f(x);
    if (key in result) {
      result[key].push(x);
    } else {
      result[key] = [x];
    }
  }
  return result;
});
exports.splitAt = curry(function(n, xs){
  if (n < 0) {
    n = 0;
  }
  return [xs.slice(0, n), xs.slice(n)];
});
exports.index = curry(function(elem, xs){
  var i$, len$, i, x;
  for (i$ = 0, len$ = xs.length; i$ < len$; ++i$) {
    i = i$;
    x = xs[i$];
    if (x === elem) {
      return i;
    }
  }
});
exports.indices = curry(function(elem, xs){
  var i$, len$, i, x, results$ = [];
  for (i$ = 0, len$ = xs.length; i$ < len$; ++i$) {
    i = i$;
    x = xs[i$];
    if (x === elem) {
      results$.push(i);
    }
  }
  return results$;
});
exports.findIndex = curry(function(f, xs){
  var i$, len$, i, x;
  for (i$ = 0, len$ = xs.length; i$ < len$; ++i$) {
    i = i$;
    x = xs[i$];
    if (f(x)) {
      return i;
    }
  }
});
exports.findIndices = curry(function(f, xs){
  var i$, len$, i, x, results$ = [];
  for (i$ = 0, len$ = xs.length; i$ < len$; ++i$) {
    i = i$;
    x = xs[i$];
    if (f(x)) {
      results$.push(i);
    }
  }
  return results$;
});
function in$(x, xs){
  var i = -1, l = xs.length >>> 0;
  while (++i < l) if (x === xs[i]) return true;
  return false;
}
